harness/determined/estimator/_reducer.py

```diff
@@ -0,0 +1,168 @@
+        cross_slot_reduce.
+
+        In the example of the calculating a distributed mean, accumulate might keep a running sum
+        and a count of metrics received:
+
+        .. code:: python
+
+            def accumulate(self, metric):
+                self.sum += metric
+                self.count += 1
+                return self.sum, self.count
+        """
+        pass
+
+    @abc.abstractmethod
+    def cross_slot_reduce(self, per_slot_metrics: List[Any]) -> Any:
+        """
+        cross_slot_reduce is called on the list of results from the final call to accumulate on
+        each slot.  per_slot_metrics will be a list of length N, where N is the number of slots in
+        the trial (or 1 in non-distributed training).  cross_slot_reduce must return the final
+        metric.
+
+        In the example of calculating a distributed mean, cross_slot_reduce might recieve a list of
+        (sum, count) tuples and it would calculate the overall mean.
+
+        .. code:: python
+
+            def cross_slot_reduce(self, per_slot_metrics):
+                sums, counts = zip(*per_slot_metrics)
+                return np.array(sum(sums) / sum(counts))
+        """
+        pass
+
+
+class _SimpleMetricReducer(MetricReducer):
+    """_SimpleMetricReducer takes a one-step reducer function and converts it to a MetricReducer."""
+
+    def __init__(self, reduce_fn: Callable[[List[Any]], Any]):
+        self.updates = []  # type: List[Any]
+        self.reduce_fn = reduce_fn
+
+    def reset(self) -> None:
+        self.updates = []
+
+    def accumulate(self, metric: Any) -> List[Any]:
+        self.updates.append(metric)
+        return self.updates
+
+    def cross_slot_reduce(self, per_slot_metrics: List[List[Any]]) -> Any:
+        flat_metrics = [item for sublist in per_slot_metrics for item in sublist]
+        return self.reduce_fn(flat_metrics)
+
+
+class _DistributedMetric(tf.keras.metrics.Metric):  # type: ignore
+    """
+    _DistributedMetric is a tf.keras Metric which should be compatible with tf.estimators and
+    which uses a MetricReducer to calculate the metric.
+    """
+
+    def __init__(
+        self,
+        context: estimator.EstimatorExperimentalContext,
+        metric: Any,
+        reducer: MetricReducer,
+        numpy_dtype: Any,
+        name: str,
+    ) -> None:
+        super().__init__(name=name)
+        # Don't conflict with tf.keras.metrics.Metric names.
+        self._det_context = context
+        self._det_reducer = reducer
+        self._det_last_accumulate = None
+
+        if isinstance(numpy_dtype, tf.dtypes.DType):
+            raise TypeError(f"numpy_dtype parameter must not be a TensorFlow dtype: {numpy_dtype}")
+        self._det_np_dtype = np.dtype(numpy_dtype)
+        self._det_tf_dtype = tf.compat.v1.as_dtype(numpy_dtype)
+
+        self.update_state(metric)
+
+        @self._det_context._build_allgather_op
```

Comment:
question: why make this a decorator and not just a function call?

Reply:
Talked offline, the question was about calling `_build_allgather_op` directly rather than applying it as a decorator.  I don't care either way, so I took the direct call approach.